### PR TITLE
Akka Usage Cleanup for Actor State safety

### PR DESCRIPTION
**Pull request by bwmcadams**
( Relates to enhancement issue I opened #2803 )
Some suggested cleanups of Akka Actors to improve the safety of Actors' internal state. Should make the actor usage safer and be more in line with best practice.

Specifically, the use of immutable collections stored in mutable `var`s is safer than immutable `val`s with mutable collections. With an immutable variable (`val`) pointing to a mutable object (such as a collection, e.g., `Vector` or `Map`), should the variable be closed over or sent in a message accidentally, the Actor's internal state could be inadvertently mutated. With a *mutable* variable (`var`) pointing at an immutable object (collection, etc.) should this state accidentally leak in any fashion, it won't be changeable from the outside.

### CODE REVIEW

File: core/controller/src/main/scala/whisk/core/controller/actions/PrimitiveActions.scala [hunk 1]
```diff
@@ -258,7 +257,7 @@ protected[actions] object ActivationFinisher {
     // when the future completes, self-destruct
     promise.future.andThen { case _ => shutdown() }
 
-    val preemptiveMsgs: Buffer[Cancellable] = Buffer.empty
+    var preemptiveMsgs = Vector.empty[Cancellable]
```
Comment (markusthoemmes) [inline core/controller/src/main/scala/whisk/core/controller/actions/PrimitiveActions.scala]: Mega nit: Should this be an `immutable.Buffer` for its constant append time?

File: core/invoker/src/main/scala/whisk/core/containerpool/ContainerPool.scala [hunk 2]
```diff
@@ -105,8 +105,8 @@ class ContainerPool(childFactory: ActorRefFactory => ActorRef,
 
       container match {
         case Some((actor, data)) =>
-          busyPool.update(actor, data)
-          freePool.remove(actor)
+          busyPool = busyPool + (actor → data)
```
Comment (markusthoemmes) [inline core/invoker/src/main/scala/whisk/core/containerpool/ContainerPool.scala]: We don't use ASCII arrows in our codebase.

Comment (bwmcadams) [reply core/invoker/src/main/scala/whisk/core/containerpool/ContainerPool.scala]: Fixing; was the default setup for my editor. longterm, having a `scalafmt` style is probably a good idea. I know IntelliJ can import it, suspect Eclipse, etc can as well.

File: core/controller/src/main/scala/whisk/core/controller/actions/PrimitiveActions.scala (hunk 1 above)
Comment (bwmcadams) [reply core/controller/src/main/scala/whisk/core/controller/actions/PrimitiveActions.scala]: There isn't an Immutable Buffer in the Scala collection library. 

[`Vector` is effectively constant on all operations](http://docs.scala-lang.org/overviews/collections/performance-characteristics.html) and the best alternative in the immutable side.

Comment (markusthoemmes) [reply core/controller/src/main/scala/whisk/core/controller/actions/PrimitiveActions.scala]: Gnah, I of course meant the immutable.Queue. My bad.

Comment (bwmcadams) [reply core/controller/src/main/scala/whisk/core/controller/actions/PrimitiveActions.scala]: You're right; we are only appending to the datastructure. `Queue`'s append is constant time versus `Vector`s which is effectively constant.

Comment (bwmcadams) [reply core/controller/src/main/scala/whisk/core/controller/actions/PrimitiveActions.scala]: As you say, performance issue isn't there so I'd probably still go with a `Vector` personally.

File: common/scala/src/main/scala/whisk/core/connector/MessageConsumer.scala [hunk 3]
```diff
@@ -202,7 +206,11 @@ class MessageFeed(description: String,
   private def sendOutstandingMessages(): Unit = {
     val occupancy = outstandingMessages.size
     if (occupancy > 0 && handlerCapacity > 0) {
-      val (topic, partition, offset, bytes) = outstandingMessages.dequeue()
+      // Easiest way with an immutable queue to cleanly dequeue
+      // Head is the first elemeent of the queue, desugared w/ a assignment pattern
```
Comment (rabbah) [inline common/scala/src/main/scala/whisk/core/connector/MessageConsumer.scala]: a assignment -> an assignment

**Comment by rabbah** [comment]
+1: Thanks for your first contribution and welcome on board 🎉

File: common/scala/src/main/scala/whisk/core/connector/MessageConsumer.scala (hunk 3 above)
Comment (bwmcadams) [reply common/scala/src/main/scala/whisk/core/connector/MessageConsumer.scala]: ? Not sure what you meant here.

Comment (rabbah) [reply common/scala/src/main/scala/whisk/core/connector/MessageConsumer.scala]: typo: desugared w/ a assignment -> desugared w/ an assignment

**Comment by bwmcadams** [comment]
I'm having trouble figuring out why the build is failing here? The only error I see is a compiler timeout.

**Comment by rabbah** [comment]
```
> travis 279666212
whisk.core.cli.test.WskEntitlementTests > Wsk Package Action should get and invoke an action from package FAILED
whisk.core.cli.test.WskEntitlementTests > Wsk Package Action should invoke an action sequence from package FAILED
whisk.core.cli.test.WskEntitlementTests > Wsk Package Action should invoke a packaged action not owned by the subject to get the subject's namespace FAILED
whisk.core.cli.test.WskEntitlementTests > Wsk Trigger Feed should not create a trigger with timeout error when feed fails to initialize FAILED
whisk.core.cli.test.WskBasicUsageTests > Wsk CLI usage should allow a 3 part Fully Qualified Name (FQN) without a leading '/' FAILED
whisk.core.cli.test.WskBasicUsageTests > Wsk actions should invoke an action that exits during initialization and get appropriate error FAILED
whisk.core.cli.test.WskBasicUsageTests > Wsk actions should invoke an action that hangs during initialization and get appropriate error FAILED
whisk.core.cli.test.WskBasicUsageTests > Wsk actions should invoke an action that exits during run and get appropriate error FAILED
whisk.core.cli.test.WskBasicUsageTests > Wsk actions should retrieve the last activation using --last flag FAILED
whisk.core.cli.test.WskBasicUsageTests > Wsk actions should ensure timestamp and stream are stripped from log lines FAILED
whisk.core.cli.test.WskBasicUsageTests > Wsk actions should ensure keys are not omitted from activation record FAILED
whisk.core.cli.test.WskBasicUsageTests > Wsk actions should write the action-path and the limits to the annotations FAILED
whisk.core.cli.test.WskBasicUsageTests > Wsk actions should create, and invoke an action that utilizes an invalid docker container with appropriate error FAILED
whisk.core.cli.test.WskBasicUsageTests > Wsk actions should invoke an action using npm openwhisk FAILED
whisk.core.cli.test.WskBasicUsageTests > Wsk actions should invoke an action receiving context properties FAILED
whisk.core.cli.test.WskBasicUsageTests > Wsk actions should invoke an action successfully with options --blocking and --result FAILED
whisk.core.cli.test.WskBasicUsageTests > Wsk actions should invoke an action that returns a result by the deadline FAILED
whisk.core.cli.test.WskBasicUsageTests > Wsk actions should invoke an action twice, where the first times out but the second does not and should succeed FAILED
whisk.core.cli.test.WskBasicUsageTests > Wsk actions should invoke action while not encoding &, <, > characters FAILED
whisk.core.cli.test.WskBasicUsageTests > Wsk actions should limit length of HTTP request and response bodies for --verbose FAILED
```

File: core/invoker/src/main/scala/whisk/core/containerpool/ContainerPool.scala [hunk 4]
```diff
@@ -115,26 +115,31 @@ class ContainerPool(childFactory: ActorRefFactory => ActorRef,
 
     // Container is free to take more work
     case NeedWork(data: WarmedData) =>
-      freePool.update(sender(), data)
-      busyPool.remove(sender()).foreach(_ => feed ! MessageFeed.Processed)
+      freePool = freePool + (sender() -> data)
+      busyPool = busyPool - sender()
+      busyPool.foreach { _ =>
```
Comment (markusthoemmes) [inline core/invoker/src/main/scala/whisk/core/containerpool/ContainerPool.scala]: The logic here is wrong. Above code said: "remove this element, and if it was actually there, also send a message". So it's going to be more like:

```scala
if(busyPool.contains(sender()) {
  busyPool = busyPool - sender()
  feed ! MessageFeed.Processed
}
```

File: core/invoker/src/main/scala/whisk/core/containerpool/ContainerPool.scala [hunk 5]
```diff
@@ -115,26 +115,31 @@ class ContainerPool(childFactory: ActorRefFactory => ActorRef,
...
 
     // Container got removed
     case ContainerRemoved =>
-      freePool.remove(sender())
-      busyPool.remove(sender()).foreach(_ => feed ! MessageFeed.Processed)
+      freePool = freePool - sender()
+      busyPool = busyPool - sender()
+      busyPool.foreach { _ =>
+        feed ! MessageFeed.Processed
+      }
```
Comment (markusthoemmes) [inline core/invoker/src/main/scala/whisk/core/containerpool/ContainerPool.scala]: Same as above.

File: core/invoker/src/main/scala/whisk/core/containerpool/ContainerPool.scala (hunk 4 above)
Comment (bwmcadams) [reply core/invoker/src/main/scala/whisk/core/containerpool/ContainerPool.scala]: oof. that was a stupid bug, but explains the behavior I was seeing in my failed tests.

**Comment by bwmcadams** [comment]
Looks like the bugs are resolved and build passes, if anyone is up for revisiting this PR.

**Comment by rabbah** [comment]
PG2/2258 🚀

**Pull request merged**